lib/charms/mongodb/v0/mongos.py

```diff
@@ -0,0 +1,168 @@
+        # Auth DB should be specified while user connects to application DB.
+        auth_source = ""
+        if self.database != "admin":
+            auth_source = "&authSource=admin"
+        return (
+            f"mongodb://{quote_plus(self.username)}:"
+            f"{quote_plus(self.password)}@"
+            f"{hosts}/{quote_plus(self.database)}?"
+            f"{auth_source}"
+        )
+
+
+class NotReadyError(PyMongoError):
+    """Raised when not all replica set members healthy or finished initial sync."""
+
+
+class MongosConnection:
+    """In this class we create connection object to Mongos.
+
+    Real connection is created on the first call to Mongos.
+    Delayed connectivity allows to firstly check database readiness
+    and reuse the same connection for an actual query later in the code.
+
+    Connection is automatically closed when object destroyed.
+    Automatic close allows to have more clean code.
+
+    Note that connection when used may lead to the following pymongo errors: ConfigurationError,
+    ConfigurationError, OperationFailure. It is suggested that the following pattern be adopted
+    when using MongoDBConnection:
+
+    with MongoMongos(self._mongos_config) as mongo:
+        try:
+            mongo.<some operation from this class>
+        except ConfigurationError, ConfigurationError, OperationFailure:
+            <error handling as needed>
+    """
+
+    def __init__(self, config: MongosConfiguration, uri=None, direct=False):
+        """A MongoDB client interface.
+
+        Args:
+            config: MongoDB Configuration object.
+            uri: allow using custom MongoDB URI, needed for replSet init.
+            direct: force a direct connection to a specific host, avoiding
+                    reading replica set configuration and reconnection.
+        """
+        self.mongodb_config = config
+
+        if uri is None:
+            uri = config.uri
+
+        self.client = MongoClient(
+            uri,
+            directConnection=direct,
+            connect=False,
+            serverSelectionTimeoutMS=1000,
+            connectTimeoutMS=2000,
+        )
+        return
+
+    def __enter__(self):
+        """Return a reference to the new connection."""
+        return self
+
+    def __exit__(self, object_type, value, traceback):
+        """Disconnect from MongoDB client."""
+        self.client.close()
+        self.client = None
+
+    def get_shard_members(self) -> Set[str]:
+        """Gets shard members.
+
+        Returns:
+            A set of the shard members as reported by mongos.
+
+        Raises:
+            ConfigurationError, ConfigurationError, OperationFailure
+        """
+        shard_list = self.client.admin.command("listShards")
+        curr_members = [
+            self._hostname_from_hostport(member["host"]) for member in shard_list["shards"]
+        ]
+        return set(curr_members)
+
+    def add_shard(self, shard_name, shard_hosts, shard_port=27017):
```

Comment:
Could we use a constant from config here? :)

Reply:
good idea, thank you